### PR TITLE
Cbqi bv quick heuristics

### PR DESCRIPTION
**Pull request by ajreynol**
Adds two heuristics for cbqi-bv, both disabled by default.
- The first optimistically solves for boundary points of inequalities.
- The second randomly interleaves inversion and value instantiations.

Adds some newly solved regressions from SMT LIB.

### CODE REVIEW

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 1]
```diff
@@ -934,19 +945,59 @@ Node BvInstantiator::hasProcessAssertion(CegInstantiator* ci, SolvedForm& sf,
...
+        if (k == BITVECTOR_ULT) {
+          k = BITVECTOR_UGE;
+        } else if (k == BITVECTOR_ULE) {
+          k = BITVECTOR_UGT;
+        } else if (k == BITVECTOR_SLT) {
+          k = BITVECTOR_SGE;
+        } else {
+          Assert(k == BITVECTOR_SLE);
+          k = BITVECTOR_SGT;
+        }
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: SLT and SLE should be rewritten to ULT and ULE. Maybe Assert that (k == BITVECTOR_ULT && k == BITVECTOR_ULE) to make sure we don't miss anything and remove the SLT and SLE cases?

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 2]
```diff
@@ -934,19 +945,59 @@ Node BvInstantiator::hasProcessAssertion(CegInstantiator* ci, SolvedForm& sf,
...
+      // point, for example
+      //   for s >= t, we solve s = t
+      //   for s > t, we solve s = t+1
+      //   for ~( s >= t ), we solve s+1 = t
+      unsigned one = 1;
+      BitVector bval(atom[0].getType().getConst<BitVectorSize>(), one);
+      Node bv_one = NodeManager::currentNM()->mkConst<BitVector>(bval);
+      Node ret;
+      if (k == BITVECTOR_ULE || k == BITVECTOR_SLE || k == BITVECTOR_UGE ||
+          k == BITVECTOR_SGE) {
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: Then we can also remove the signed ops here.

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 3]
```diff
@@ -934,19 +945,59 @@ Node BvInstantiator::hasProcessAssertion(CegInstantiator* ci, SolvedForm& sf,
...
+        ret = atom[0].eqNode(atom[1]);
+      } else if (k == BITVECTOR_ULT || k == BITVECTOR_SLT) {
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: And here.

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 4]
```diff
@@ -934,19 +945,59 @@ Node BvInstantiator::hasProcessAssertion(CegInstantiator* ci, SolvedForm& sf,
...
+        ret = NodeManager::currentNM()
+                  ->mkNode(kind::BITVECTOR_PLUS, atom[0], bv_one)
+                  .eqNode(atom[1]);
+      } else {
+        Assert(k == BITVECTOR_UGT || k == BITVECTOR_SGT);
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: And here.

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 5]
```diff
@@ -976,79 +1027,92 @@ bool BvInstantiator::processAssertions(CegInstantiator* ci, SolvedForm& sf,
...
+      // get inst id list
+      Trace("cegqi-bv") << "  " << iti->second.size()
+                        << " candidate instantiations for " << pv << " : "
+                        << std::endl;
+      // the order of instantiation ids we will try
+      std::vector<unsigned> inst_ids_try;
+      // until we have a model-preserving selection function for BV, this must
+      // be heuristic (we only pick one literal)
+      // hence we randomize the list
+      // this helps robustness, since picking the same literal every time may be
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: "may be lead" -> "may lead" ?

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 6]
```diff
@@ -976,79 +1027,92 @@ bool BvInstantiator::processAssertions(CegInstantiator* ci, SolvedForm& sf,
...
+          if (!curr_slack_val.isNull()) {
+            Trace("cegqi-bv") << "   ...with slack value : " << curr_slack_val
+                              << std::endl;
           }
+          // process information about solved status
+          std::unordered_map<unsigned, BvInverterStatus>::iterator its =
+              d_inst_id_to_status.find(inst_id);
+          Assert(its != d_inst_id_to_status.end());
+          if (!(*its).second.d_conds.empty()) {
+            Trace("cegqi-bv") << "   ...with " << (*its).second.d_conds.size()
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: (*its).second -> its->second

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 7]
```diff
@@ -976,79 +1027,92 @@ bool BvInstantiator::processAssertions(CegInstantiator* ci, SolvedForm& sf,
...
+                              << " side conditions : " << std::endl;
+            for (unsigned k = 0; k < (*its).second.d_conds.size(); k++) {
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: ditto

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 8]
```diff
@@ -976,79 +1027,92 @@ bool BvInstantiator::processAssertions(CegInstantiator* ci, SolvedForm& sf,
...
+              Node cond = (*its).second.d_conds[k];
```
Comment (aniemetz) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: ditto

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 9]
```diff
@@ -920,11 +923,23 @@ Node BvInstantiator::hasProcessAssertion(CegInstantiator* ci, SolvedForm& sf,
...
+    if (k == EQUAL) {
+      // always use slack for disequalities
+      useSlack = true;
+    } else if( k == BITVECTOR_ULT || k == BITVECTOR_SLT ){
+      if (options::cbqiBvSlackIneq()) {
+        useSlack = true;
+      }
+    }else{
+      // others should be rewritten
+      Assert( false );
```
Comment (mpreiner) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: Isn't here Unreachable() better?

File: src/theory/quantifiers/ceg_t_instantiator.cpp [hunk 10]
```diff
@@ -976,79 +1015,94 @@ bool BvInstantiator::processAssertions(CegInstantiator* ci, SolvedForm& sf,
...
-
-      // currently we take select the first literal
-      if( inst_ids_try.empty() ){
-        // try the first one
-        inst_ids_try.push_back( inst_id );
-      } else {
-        // selection criteria across multiple literals goes here
-
 
+        // currently we take select the first literal
```
Comment (mpreiner) [inline src/theory/quantifiers/ceg_t_instantiator.cpp]: Type 'take select'?

**Pull request merged**